src/lib/shell/commands/BDXDownloader_nrfconnect.h

```diff
@@ -0,0 +1,130 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <lib/support/TypeTraits.h>
+#include <lib/support/logging/CHIPLogging.h>
+#include <messaging/ExchangeContext.h>
+#include <protocols/bdx/BdxMessages.h>
+#include <protocols/bdx/BdxTransferSession.h>
+#include <protocols/bdx/TransferFacilitator.h>
+
+#include <dfu/dfu_target.h>
+#include <dfu/dfu_target_mcuboot.h>
+
+namespace chip {
+namespace Shell {
+
+class BdxDownloader : public bdx::Initiator
+{
+public:
+    void SetInitialExchange(Messaging::ExchangeContext * ec) { mExchangeCtx = ec; }
+
+private:
+    void HandleTransferSessionOutput(bdx::TransferSession::OutputEvent & event) override
+    {
+        using OutputEventType  = bdx::TransferSession::OutputEventType;
+        using MessageType      = bdx::MessageType;
+        using SendMessageFlags = Messaging::SendMessageFlags;
+
+        CHIP_ERROR error = CHIP_NO_ERROR;
+        int status       = 0;
+
+        switch (event.EventType)
+        {
+        case OutputEventType::kNone:
+            if (mIsTransferComplete)
+            {
+                ChipLogProgress(BDX, "Transfer complete!");
+                mTransfer.Reset();
+                mIsTransferComplete = false;
+                status              = dfu_target_done(true);
```

Comment:
Does this apply the image? Applying the image automatically is technically not to spec. When the download is complete you should call `ApplyUpdateRequest` and then apply the image based on `ApplyUpdateResponse`.

Reply:
Right, added two more commands: "show-update" for displaying the token, and "apply" for sending ApplyUpdate request.